### PR TITLE
Assignment's Due Date

### LINKED ISSUE #9: [WORKSPACE] Add due date for assignment
Workspace owners have the ability to configure the due dates of an assignment. Assignments submitted by workspace participants are then marked with either an `in-time` or `late` flag for score calculation.

### PR DESCRIPTION
**Pull request by jate-koh**
### What
Briefly describe this feature / fix / refactor

- In short, this branch covers the following features:
   - Adding Due Date: 16f0ff6
   - Adding Late Flag and its Logic: e49fb60
   - Changing Database Schema and Migration Script: 5b09fc6, 221c4e9

### Why
Describe the reason why we need this feature / fix / refactor

- **Obviously, this is an essential feature**. While this software is a programming grader and not LEB2, it's still a platform for students to submit their work. Assignment's deadline, due date, and such must exist or be implemented in order to encourage students to devote their effort to class and not slack off by driving Tesla around the country.

### How
Describe the feature / fix / refactor in detail.

- Explain the technical solution you have done to create / fix this pull request
   - Added **Due Date** field in the assignment domain as with the `nullable` pointer data type of timestamp (or 
    `*time.Time`). The change can be observed in 16f0ff6.
   - **Due Date** is also included in `assignment` table as field named `due_date` in the new database schema. It's written in complete version of migration script in 221c4e9.
   - **Is Late** boolean flag is also added to detect whether that student submission is sent on-time or not. The logic behind its can be view fully in e49fb60. but here's snippet of the logic to determine the lateness flag:
```go
if assignment.DueDate == nil {
     submission.IsLate = false
} else if submission.SubmittedAt.After(*assignment.DueDate) {
     submission.IsLate = true
} else {
     submission.IsLate = false
}
```
> In a nutshell, late flag will be set false if either that submission in sent **before** the assignment due date or the assignment's due date **does not exist**. Late flag will become true if the submission is sent **after** the assignment's due date.

- Attach the screenshot (if applicable)

Screenshot 2566-11-16 at 00 18 48

>  Due date field is shown when getting API for `/workspaces/<workspace_id>/assignments`

Screenshot 2566-11-16 at 22 03 27

> Upon getting API in `/workspaces/<workspace_id>/assignments/<assign_id>/submissions`, the late flag is shown false because these submissions are submitted before the set due date.

Screenshot 2566-11-16 at 22 03 17

> Vice versa, the late flag is true since these submissions are submitted after the due date.

### Todo list (if applicable)
- [x] Open my computer
- [x] Add due date field.
- [x] Make change to database schema and migration files.
- [x] Add late flag to which its value is determined by assignment's due date **when listing user's submissions**.
- [ ] Add late flag to which its value is determined by assignment's due date **when listing user's assignments**.

## Checklist
- [x] Explained the purpose of this PR 
- [x] Tested on **Local machine** and verified that there are no visible errors
- [ ] Tested on **Staging server** and verified that there are no visible errors
- [ ] Get enough sleep 💤
- [ ] Updated documentation (if applicable)
- [ ] Added or updated test suite for the changes (if applicable)

### CODE REVIEW

**Comment by XiaoXuxxxx** [comment]
ตึงเกิ้นๆๆๆๆ

File: repository/assignment.go [hunk 1]
```diff
@@ -214,9 +214,25 @@ func (r *assignmentRepository) ListSubmission(userId string, assignmentId int) (
...
 		for i := range submissions {
 			submissionById[submissions[i].Id] = &submissions[i]
+			submission := submissionById[submissions[i].Id]
+			if assignment.DueDate == nil {
+				submission.IsLate = false
+			} else if submission.SubmittedAt.After(*assignment.DueDate) {
+				submission.IsLate = true
+			} else {
+				submission.IsLate = false
+			}
```
Comment (riflowth) [inline repository/assignment.go:233]: You can merge the condition branches that give the same result together.

Comment (riflowth) [inline repository/assignment.go]: This aggregation could be done in SQL to avoid extra queries.

**Review by riflowth — CHANGES_REQUESTED** [review]

**Comment by riflowth** [comment]
> This pull request is a masterpiece of detailed writing, setting a great example for other developers as well.
kudos to @jate-koh 👍

File: repository/assignment.go (hunk 1 above)
Comment (jate-koh) [reply repository/assignment.go]: Acknowledged 🫡
Working on it~

**Comment by jate-koh** [comment]
@riflowth, in the latest commit a4e1fdd, I have removed the unnecessary querying of the assignment table in line 217.
```go
assignment, err := r.Get(assignmentId, userId)
```

I added `due_date` field to `submission` domain.
```go
type Submission struct {
        // ... same field
	// Always aggregation
	DueDate *time.Time         `json:"dueDate,omitempty" db:"due_date"`
	IsLate  bool               `json:"isLate"`
	Results []SubmissionResult `json:"results,omitempty"`
}
```

Then adjust the `submission` query to `RIGHT JOIN` with `assignment` to acquire the `due_date` field.
```go
submissions := make([]domain.Submission, 0)
err := r.db.Select(
        &submissions,
       `SELECT s.*, a.due_date
         FROM submission s
         RIGHT JOIN assignment a ON a.id = s.assignment_id
         WHERE s.assignment_id = ? AND s.user_id = ?`,
         assignmentId,
         userId,
)
if err != nil {
         return nil, fmt.Errorf("cannot query to list submission: %w", err)
}
```

This is the current result when GET-ing submissions via API in `/workspaces/<workspace_id>/assignments/<assign_id>/submissions`
Screenshot 2566-11-21 at 14 40 46

**Comment by riflowth** [comment]
@jate-koh According to our onsite-discussion, we don't need `dueDate` field when getting submission. Could you please remove it?

**Comment by jate-koh** [comment]
@riflowth, I have done the tasks as you requested. The change can be viewed via dd7bb74.

Mainly, I have removed the `dueDate` field from the API payload.
Screenshot 2566-11-21 at 21 36 16

I removed `dueDate` from the `submission` domain to achieve that part. Then, I rewrote the query to determine the `is_late` flag in the querying process. Here's the new query:
```go
query := `
  SELECT
    s.*,
    CASE
      WHEN s.submitted_at > a.due_date THEN TRUE
      WHEN s.submitted_at < a.due_date THEN FALSE
      WHEN s.submitted_at = a.due_date THEN FALSE
      ELSE FALSE
    END AS is_late
  FROM submission s
  INNER JOIN assignment a on a.id = s.assignment_id
  WHERE s.assignment_id = ? AND s.user_id = ?
`
err := r.db.Select(&submissions, query, assignmentId, userId)
if err != nil {
return nil, fmt.Errorf("cannot query to list submission: %w", err)
}
```

Do not hesitate to contact me back again, if you want me to rewrite or fix something 😉

**Review by riflowth — APPROVED** [review]
**Nicely done!** Thanks for your work 🙏

**Pull request merged**